Mathlib/Data/Set/Ncard.lean

```diff
@@ -34,20 +43,402 @@ in `Set α`) via default arguments.
+  · obtain ⟨t₀, ht₀s, ht₀⟩ := IH (le_trans (by simp) hle)
+    simp only [Nat.cast_succ] at *
+    have hne : t₀ ≠ s
+    · rintro rfl; rw [ht₀, ←Nat.cast_one, ←Nat.cast_add, Nat.cast_le] at hle; simp at hle
+    obtain ⟨x, hx⟩ := exists_of_ssubset (ht₀s.ssubset_of_ne hne)
+    exact ⟨insert x t₀, insert_subset hx.1 ht₀s, by rw [encard_insert_of_not_mem hx.2, ht₀]⟩
+  simp only [top_le_iff, encard_eq_top_iff]
+  exact fun _ hi ↦ ⟨s, Subset.rfl, hi⟩
+
+theorem exists_supset_subset_encard_eq (hst : s ⊆ t) (hsk : s.encard ≤ k) (hkt : k ≤ t.encard) :
+    ∃ r, s ⊆ r ∧ r ⊆ t ∧ r.encard = k := by
+  obtain (hs | hs) := eq_or_ne s.encard ⊤
+  · rw [hs, top_le_iff] at hsk; subst hsk; exact ⟨s, Subset.rfl, hst, hs⟩
+  obtain ⟨k, rfl⟩ := exists_add_of_le hsk
+  obtain ⟨k', hk'⟩ := exists_add_of_le hkt
+  have hk : k ≤ encard (t \ s)
+  · rw [←encard_diff_add_encard_of_subset hst, add_comm] at hkt
+    exact WithTop.le_of_add_le_add_right hs hkt
+  obtain ⟨r', hr', rfl⟩ := exists_subset_encard_eq hk
+  refine' ⟨s ∪ r', subset_union_left _ _, union_subset hst (hr'.trans (diff_subset _ _)), _⟩
+  rw [encard_union_eq (disjoint_of_subset_right hr' disjoint_sdiff_right)]
+
+section Function
+
+variable {s : Set α} {t : Set β} {f : α → β}
+
+theorem encard_image_of_injOn (h : InjOn f s) : (f '' s).encard = s.encard := by
+  rw [encard, PartENat.card_image_of_injOn h, encard]
+
+theorem encard_congr (e : s ≃ t) : s.encard = t.encard := by
+  rw [←encard_univ_coe, ←encard_univ_coe t, encard_univ, encard_univ, PartENat.card_congr e]
+
+theorem encard_image_of_injective (hf : f.Injective) (s : Set α) : (f '' s).encard = s.encard :=
+  encard_image_of_injOn (hf.injOn s)
+
+theorem encard_le_of_embedding (e : s ↪ t) : s.encard ≤ t.encard := by
+  rw [←encard_univ_coe, ←encard_image_of_injective e.injective univ,
+    ←encard_image_of_injective Subtype.coe_injective]
+  exact encard_mono (by simp)
+
+theorem encard_image_le (f : α → β) (s : Set α) : (f '' s).encard ≤ s.encard := by
+  obtain (h | h) := isEmpty_or_nonempty α
+  · rw [s.eq_empty_of_isEmpty]; simp
+  rw [←encard_image_of_injOn (invFunOn_injOn_image f s)]
+  apply encard_le_of_subset
+  exact invFunOn_image_image_subset f s
+
+theorem Finite.injOn_of_encard_image_eq (hs : s.Finite) (h : (f '' s).encard = s.encard) :
+    InjOn f s := by
+  obtain (h' | hne) := isEmpty_or_nonempty α
+  · rw [s.eq_empty_of_isEmpty]; simp
+  rw [←encard_image_of_injOn (invFunOn_injOn_image f s)] at h
+  rw [injOn_iff_invFunOn_image_image_eq_self]
+  exact hs.eq_of_subset_of_encard_le (invFunOn_image_image_subset f s) h.symm.le
+
+theorem encard_preimage_of_injective_subset_range (hf : f.Injective) (ht : t ⊆ range f) :
+    (f ⁻¹' t).encard = t.encard := by
+  rw [←encard_image_of_injective hf, image_preimage_eq_inter_range,
+    inter_eq_self_of_subset_left ht]
+
+theorem encard_le_encard_of_injOn (hf : ∀ a ∈ s, f a ∈ t) (f_inj : InjOn f s) :
```

Comment:
`hf : MapsTo f s t`, and you probably want to rename this to take advantage of dot notation.